qiskit/extensions/quantum_initializer/diagonal.py

```diff
@@ -46,28 +47,22 @@ def __instancecheck__(mcs, inst):
 
 class DiagonalGate(Gate, metaclass=DiagonalMeta):
     """
-    diag =  list of the 2^k diagonal entries (for a diagonal gate on k qubits). Must contain at
-    least two entries.
+    diag (list, ndarray): iterable of the 2^k diagonal entries (for a diagonal gate on k qubits).
+       Must contain at least two entries.
     """
 
     def __init__(self, diag):
         """Check types"""
-        # Check if diag has type "list"
-        if not isinstance(diag, list):
-            raise QiskitError("The diagonal entries are not provided in a list.")
-        # Check if the right number of diagonal entries is provided and if the diagonal entries
-        # have absolute value one.
+        try:
+            diag = np.array(diag, dtype=complex)
+        except TypeError:
+            raise QiskitError("Not all of the diagonal entries can be converted to "
+                              "complex numbers.")
         num_action_qubits = math.log2(len(diag))
         if num_action_qubits < 1 or not num_action_qubits.is_integer():
             raise QiskitError("The number of diagonal entries is not a positive power of 2.")
-        for z in diag:
-            try:
-                complex(z)
-            except TypeError:
```

Comment:
This is not tested anymore, but I cannot think of a case where this test would've failed in the first place... Do you have an idea why this existed?

Reply:
This try statement was converted to numpy array conversion above so it essentially still exists.